### PR TITLE
Misc fixes

### LINKED ISSUE #9921: Determine how to implement retries
Do it in the method body like EH or have an abstraction to handle it
https://github.com/Azure/azure-sdk-for-net/pull/9842#discussion_r377117576

### PR DESCRIPTION
**Pull request by JoshLove-msft**
- Update client hierarchy based on our discussions
  - 4 clients with 2 abstract base classes
  - remove Receiver/Sender namespaces 
- Add RunOperation to RetryPolicy to help simplify retry logic
- Remove SessionClient and put all session operations under new Session property on Receiver
- Include SessionOptions in ctor overloads
- Set the session filter when using a sessionful entity even if no session Id is populated

Fixes https://github.com/Azure/azure-sdk-for-net/issues/9921

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/Directory.Build.props [hunk 1]
```diff
@@ -3,4 +3,16 @@
     Add any shared properties you want for the projects under this package directory that need to be set before the auto imported Directory.Build.props
   -->
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., Directory.Build.props))\Directory.Build.props" />
+  
+  <PropertyGroup>
+    <!--
+    These analyzers are blocked by https://github.com/Azure/azure-sdk-tools/issues/127
+     -->
+    <NoWarn>
+      $(NoWarn);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/Directory.Build.props]: This allows us to not match the ClientOption type names exactly with the Client type.
This is useful for SenderClientOptions.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/BasicRetryPolicy.cs [hunk 2]
```diff
@@ -30,7 +30,7 @@ internal class BasicRetryPolicy : ServiceBusRetryPolicy
         ///   behavior.
         /// </summary>
         ///
-        public ServiceBusRetryOptions Options { get; }
+        public override ServiceBusRetryOptions Options { get; }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/BasicRetryPolicy.cs]: Exposing this in ServiceBusRetryPolicy to provide access to the default timeout.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConsumer.cs [hunk 3]
```diff
@@ -36,14 +35,14 @@ internal abstract class TransportConsumer
         public FaultTolerantAmqpObject<ReceivingAmqpLink> ReceiveLink { get; protected set; }
 
         /// <summary>
-        ///
+        /// TODO make this generic
         /// </summary>
         public AmqpConnectionScope ConnectionScope { get; set; }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConsumer.cs]: This is used within RetryPolicy.RunOperation to tell if connection is Open.
Need to make this be TransportConnectionScope and have AmqpConnectionScope derive from it.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs [hunk 4]
```diff
@@ -72,5 +83,75 @@ public abstract class ServiceBusRetryPolicy
...
+        /// </summary>
+        /// <param name="operation"></param>
+        /// <param name="entityName"></param>
+        /// <param name="scope"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        internal async Task RunOperation(
+            Func<Task> operation,
+            string entityName,
+            AmqpConnectionScope scope,
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: Should be using TransportConnectionScope

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConsumer.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConsumer.cs]: Fixed.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: Fixed.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs [hunk 5]
```diff
@@ -23,7 +23,7 @@ public class QueueReceiverClient : ServiceBusReceiverClient
         /// </summary>
         protected QueueReceiverClient(){}
         /// <summary>
-        ///   Initializes a new instance of the <see cref="ServiceBusReceiverClient"/> class.
+        ///   Initializes a new instance of the <see cref="ServiceBusReceiver"/> class.
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: nit:  Doc comments refer to `ServiceBusReceiver` instead of `QueueReceiverClient`

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs [hunk 6]
```diff
@@ -42,8 +42,30 @@ public QueueReceiverClient(string connectionString)
...
         }
 
+
         /// <summary>
-        ///   Initializes a new instance of the <see cref="ServiceBusReceiverClient"/> class.
+        ///   Initializes a new instance of the <see cref="ServiceBusReceiver"/> class.
+        /// </summary>
+        ///
+        /// <param name="connectionString">The connection string to use for connecting to the Service Bus namespace; it is expected that the Service Bus entity name and the shared key properties are contained in this connection string.</param>
+        /// <param name="sessionOptions"></param>
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: nit:  missing text.    Leaving just this one, but all the doc comments seem to have parameters not filled in.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs [hunk 7]
```diff
@@ -42,8 +42,30 @@ public QueueReceiverClient(string connectionString)
...
+        ///
+        /// <remarks>
+        ///   If the connection string is copied from the Service Bus namespace, it will likely not contain the name of the desired Service Bus entity,
+        ///   which is needed.  In this case, the name can be added manually by adding ";EntityPath=[[ Service Bus entity NAME ]]" to the end of the
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: nit:  Capitalization in these is weird.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 8]
```diff
@@ -10,11 +10,13 @@
...
 using System.Threading.Tasks;
 using Azure.Core;
+using Azure.Messaging.ServiceBus.Amqp;
 using Azure.Messaging.ServiceBus.Core;
 using Azure.Messaging.ServiceBus.Diagnostics;
 using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Amqp.Framing;
 
-namespace Azure.Messaging.ServiceBus.Receiver
+namespace Azure.Messaging.ServiceBus.Core
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Should this type move to the `Core` folder?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs [hunk 9]
```diff
@@ -7,12 +7,12 @@
 using Azure.Core;
 using Azure.Messaging.ServiceBus.Core;
 
-namespace Azure.Messaging.ServiceBus.Receiver
+namespace Azure.Messaging.ServiceBus
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: Should these types move to the root folder as part of the namespace change?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 10]
```diff
@@ -93,93 +97,201 @@ public abstract class ServiceBusReceiverClient : IAsyncDisposable
...
+                isSessionReceiver: IsSessionReceiver);
+            Session = new ServiceBusSession(
+                Consumer,
+                RetryPolicy,
+                sessionOptions != null);
         }
 
-        /// <summary>
-        ///   Initializes a new instance of the <see cref="ServiceBusReceiverClient"/> class.
+        ///// <summary>
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Was leaving these commented intentional?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 11]
```diff
@@ -305,18 +417,25 @@ await foreach (ServiceBusMessage msg in ret.ConfigureAwait(false))
...
-            string sessionId = null,
             [EnumeratorCancellation]
             CancellationToken cancellationToken = default)
         {
-            string receiveLinkName = "";
+            if (IsSessionReceiver)
+            {
+                // if this is a session receiver, the receive link must be open in order to peek messages
+                await RetryPolicy.RunOperation(
+                    async () => await Consumer.ReceiveLink.GetOrCreateAsync(RetryPolicy.Options.TryTimeout).ConfigureAwait(false),
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: The original intent was to calculate the try timeout to allow custom retry policies to provide logic that adjusts the timeout period based on the attempt.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSession.cs [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        public virtual async Task<string> GetSessionId(CancellationToken cancellationToken = default)
+        {
+            if (!_isSessionful)
+            {
+                return null;
+            }
+            ReceivingAmqpLink openedLink = null;
+            await _retryPolicy.RunOperation(
+                async () =>
+                openedLink = await _consumer.ReceiveLink.GetOrCreateAsync(_retryPolicy.Options.TryTimeout).ConfigureAwait(false),
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSession.cs]: This confuses me.   The point of the transport consumer is to abstract the concept of AMQP links away from callers, especially those outside of the AMQP layer.   In its current form , this is leaking the abstraction.   This makes it much harder to inject a transport consumer for behavioral overrides to unit test and to potentially add other transports (like a recording framework) in the future that may not be link-based.

I could be missing something, but shouldn't we be having the details of getting the session on the transport consumer or another transport client?    I'd also advise against exposing the link directly, to avoid leaking that abstraction and force the consumer to be extended instead.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/SessionOptions.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        /// <summary>
+        ///
+        /// </summary>
+        public string SessionId { get; set; }
+
+        /// <summary>
+        ///
+        /// </summary>
+        public ServiceBusConnection Connection { get; set; }
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/SessionOptions.cs]: Maybe this was a detail from the design discussions that I missed and, if so, apologies but.... having the connection here feels orthogonal to the concerns of a session.   The connection itself isn't session-aware or  required to do something session-related?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/SubscriptionReceiverClient.cs [hunk 14]
```diff
@@ -53,16 +53,61 @@ public class SubscriptionClient : ServiceBusReceiverClient
...
-        ///   Initializes a new instance of the <see cref="ServiceBusReceiverClient"/> class.
+        ///   Initializes a new instance of the <see cref="ServiceBusReceiver"/> class.
         /// </summary>
         ///
         /// <param name="connectionString">The connection string to use for connecting to the Service Bus namespace; it is expected that the Service Bus entity name and the shared key properties are contained in this connection string.</param>
+        /// <param name="sessionOptions"></param>
+        ///
+        /// <remarks>
+        ///   If the connection string is copied from the Service Bus namespace, it will likely not contain the name of the desired Service Bus entity,
+        ///   which is needed.  In this case, the name can be added manually by adding ";EntityPath=[[ Service Bus entity NAME ]]" to the end of the
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/SubscriptionReceiverClient.cs]: nit:  The entity path capitalization is weird.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusConnection.cs [hunk 15]
```diff
@@ -27,7 +26,7 @@ namespace Azure.Messaging.ServiceBus
     ///
     /// <seealso href="https://docs.microsoft.com/en-us/Azure/event-hubs/event-hubs-about" />
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusConnection.cs]: nit:  You may want to update the reference on line 28.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusConnection.cs [hunk 16]
```diff
@@ -381,17 +385,19 @@ internal virtual TransportSender CreateTransportProducer(ServiceBusRetryPolicy r
...
             ServiceBusRetryPolicy retryPolicy,
             long? ownerLevel = default,
             uint? prefetchCount = default,
-            string sessionId = default)
+            string sessionId = default,
+            bool isSessionReceiver = default)
         {
             Argument.AssertNotNull(retryPolicy, nameof(retryPolicy));
-            return InnerClient.CreateConsumer(retryPolicy, ownerLevel, prefetchCount, sessionId);
+            return InnerClient.CreateConsumer(retryPolicy, ownerLevel, prefetchCount, sessionId, isSessionReceiver);
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusConnection.cs]: At some point, we probably want to consider unifying terminology;  if we're calling these "Sender/Receiver" rather than "Producer/Consumer"  we'll probably want to update the member names.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/SessionOptions.cs (hunk 13 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/SessionOptions.cs]: Yeah our idea was that when using sessions, you need to provide a Connection. I think this may change though so this is still pretty tentative.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs (hunk 8 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: This will probably change around again, so just leaving for now.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConnectionScope.cs [hunk 17]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.Messaging.ServiceBus.Core
+{
+    internal abstract class TransportConnectionScope : IDisposable
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConnectionScope.cs]: The reason I added this was to avoid leaking the AMQP specific AMQPConnectionScope in our RetryPolicy. RetryPolicy.RunOperation needs access to the scope to check whether it is Disposed.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportClient.cs [hunk 18]
```diff
@@ -111,14 +111,16 @@ internal abstract class TransportClient : IAsyncDisposable
...
         ///
         /// <returns>A <see cref="TransportConsumer" /> configured in the requested manner.</returns>
         ///
         public abstract TransportConsumer CreateConsumer(
             ServiceBusRetryPolicy retryPolicy,
             long? ownerLevel,
             uint? prefetchCount,
-            string sessionId);
+            string sessionId,
+            bool isSessionReceiver = default);
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportClient.cs]: Fixed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs [hunk 19]
```diff
@@ -32,6 +37,13 @@ public abstract class ServiceBusRetryPolicy
         ///
         public abstract TimeSpan CalculateTryTimeout(int attemptCount);
 
+        /// <summary>
+        ///   The set of options responsible for configuring the retry
+        ///   behavior.
+        /// </summary>
+        ///
+        public abstract ServiceBusRetryOptions Options { get; }
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: Reverted this change as we will use the per try timeout anyway.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: Fixed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: I wanted to do a separate pass on documentation before any previews to avoid everything is solid. I have issues tracking this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs (hunk 7 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: Fixed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs (hunk 9 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: This will probably change around again so leaving for now.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSession.cs (hunk 12 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSession.cs]: The implementation can be moved down to the consumer layer.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs [hunk 20]
```diff
@@ -4,6 +4,11 @@
 using System;
 using System.ComponentModel;
 using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: Yeah it probably shouldn't know about AMQP. I think the only thing that wouldn't apply generically though is this exception type, so I'm not sure that it is worth moving this to an extension method. What do you think about intoducing the concept of a TransportException  and using that here instead? AmqpException would derive from TransportException.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs (hunk 10 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Yeah those will be used if we go back to making this not be abstract -- also this should be abstract, looks like that was deleted at some point.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: I'm not opposed to a transport exception, but I'm not sure it'll work the way that you're proposing.  The `AmqpException` is part of the AMQP library and not one of ours, unfortunately.   You'd have to wrap it.  They do some specific things with how it wraps inner exceptions and such, which is why that translation method came into play, I'm not sure that we'd want that to be applied generically.... so maybe that translate could live in the AMQP area and if no other translations apply, wrap the AMQP exception in a transport exception and bubble it?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs [hunk 21]
```diff
@@ -72,5 +84,75 @@ public abstract class ServiceBusRetryPolicy
...
+            Func<Task> operation,
+            string entityName,
+            TransportConnectionScope scope,
+            CancellationToken cancellationToken)
+        {
+            var failedAttemptCount = 0;
+
+            try
+            {
+                TimeSpan tryTimeout = CalculateTryTimeout(0);
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: Yeah, if we are going to support the per-try timeout, we would need to change to Func<TimeSpan, Task>

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConnectionScope.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConnectionScope.cs]: Ah.... fair enough.   I missed that detail.   Yeah,  the only other idea that I can think of would be to pass a callback to check for disposal, and I like your approach much better.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs (hunk 6 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/QueueReceiverClient.cs]: Then I'll stop nitpicking doc comment issues.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs (hunk 11 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Fixed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: I'd like to figure this out as part of https://github.com/Azure/azure-sdk-for-net/issues/9879

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConnectionScope.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/TransportConnectionScope.cs]: Will cover as part of doc updates

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs]: Fixed

**Review by jsquire — APPROVED** [review]
